normalizer.go

```diff
@@ -0,0 +1,179 @@
+	lexer := NewSQLLexer(input)
+
+	var normalizedSQL string
+	var normalizedInfo = &NormalizedInfo{
+		Tables:   []string{},
+		Comments: []string{},
+		Commands: []string{},
+	}
+
+	var lastToken *Token // The last token that is not whitespace or comment
+
+	for token := range lexer.ScanAllTokens() {
+		if token.Type == COMMENT || token.Type == MULTILINE_COMMENT {
+			// Collect comments
+			if n.config.CollectComments {
+				normalizedInfo.Comments = append(normalizedInfo.Comments, token.Value)
+			}
+		} else if token.Type == IDENT {
+			if isCommand(strings.ToUpper(token.Value)) && n.config.CollectCommands {
+				// Collect commands
+				normalizedInfo.Commands = append(normalizedInfo.Commands, strings.ToUpper(token.Value))
+			} else if lastToken != nil && isTableIndicator(strings.ToUpper(lastToken.Value)) {
+				// Collect table names
+				if n.config.CollectTables {
+					normalizedInfo.Tables = append(normalizedInfo.Tables, token.Value)
+				}
+			}
+		}
+
+		writeNormalizedSQL(token, lastToken, &normalizedSQL)
+
+		// TODO: We rely on the WS token to determine if we should add a whitespace
+		// This is not ideal, as SQLs with slightly different formatting will NOT be normalized into single family
+		// e.g. "SELECT * FROM table where id = ?" and "SELECT * FROM table where id= ?" will be normalized into different family
+		if token.Type != WS && token.Type != COMMENT && token.Type != MULTILINE_COMMENT {
+			lastToken = token
+		}
+	}
+
+	// We use regex to group consecutive obfuscated values into single placeholder.
+	// This is "less" performant than token by token processing,
+	// but it is much simpler to implement and maintain.
+	// The trade off made here is assuming normalization runs on backend
+	// where performance is not as critical as the agent.
+	normalizedSQL = groupObfuscatedValues(normalizedSQL)
+	if !n.config.KeepSQLAlias {
+		normalizedSQL = DiscardSQLAlias(normalizedSQL)
+	}
+
+	// Dedupe collected metadata
+	dedupeNormalizedInfo(normalizedInfo)
+
+	return strings.TrimSpace(normalizedSQL), normalizedInfo, nil
+}
+
+func writeNormalizedSQL(token *Token, lastToken *Token, normalizedSQL *string) {
+	if token.Type == WS || token.Type == COMMENT || token.Type == MULTILINE_COMMENT {
+		// We don't rely on the WS token to determine if we should add a whitespace
+		return
+	}
+
+	// determine if we should add a whitespace
+	writeWhitespace(lastToken, token, normalizedSQL)
+
+	// UPPER CASE SQL keywords
+	if isSQLKeyword(token) {
+		*normalizedSQL += strings.ToUpper(token.Value)
+		return
+	}
+	*normalizedSQL += token.Value
+}
+
+// groupObfuscatedValues groups consecutive obfuscated values in a SQL query into a single placeholder.
+// It replaces "(?, ?, ...)" and "[?, ?, ...]" with "( ? )" and "[ ? ]", respectively.
+// Returns the modified SQL query as a string.
+func groupObfuscatedValues(input string) string {
+	groupable_regex := regexp.MustCompile(`(\()\s*\?(?:\s*,\s*\?\s*)*\s*(\))|(\[)\s*\?(?:\s*,\s*\?\s*)*\s*(\])`)
```

Comment:
It's probably best to move this regexp to a global variable to avoid having it be reinstantiated every time the function is called. 

```
var groupableRegex = regexp.MustCompile(`(\()\s*\?(?:\s*,\s*\?\s*)*\s*(\))|(\[)\s*\?(?:\s*,\s*\?\s*)*\s*(\])`) 
```